diff --git a/src/hotspot/share/oops/constantPool.hpp b/src/hotspot/share/oops/constantPool.hpp
index 93acd1bd2b7..afb0954e8a7 100644
--- a/src/hotspot/share/oops/constantPool.hpp
+++ b/src/hotspot/share/oops/constantPool.hpp
@@ -364,7 +364,7 @@ class ConstantPool : public Metadata {
 
   void variant_linkage_index_at_put(int which, int constant_index, int ref_index) {
     tag_at_put(which, JVM_CONSTANT_Linkage);
-    *int_at_addr(which) = ((jint) constant_index<<16) | ref_index;
+    *int_at_addr(which) = ((jint) ref_index<<16) | constant_index;
   }
 
   void unresolved_string_at_put(int which, Symbol* s) {
diff --git a/src/hotspot/share/utilities/constantTag.hpp b/src/hotspot/share/utilities/constantTag.hpp
index c3683859901..2dc22b238c4 100644
--- a/src/hotspot/share/utilities/constantTag.hpp
+++ b/src/hotspot/share/utilities/constantTag.hpp
@@ -139,8 +139,10 @@ class constantTag {
   constantTag(jbyte tag) {
     jbyte entry_tag = tag & ~JVM_CONSTANT_QDescBit;
     assert((((tag & JVM_CONSTANT_QDescBit) == 0) && (entry_tag >= 0 && entry_tag <= JVM_CONSTANT_NameAndType) ||
-           (entry_tag >= JVM_CONSTANT_MethodHandle && entry_tag <= JVM_CONSTANT_InvokeDynamic) ||
-           (entry_tag >= JVM_CONSTANT_InternalMin && entry_tag <= JVM_CONSTANT_InternalMax))
+            (entry_tag >= JVM_CONSTANT_MethodHandle && entry_tag <= JVM_CONSTANT_InvokeDynamic) ||
+            (entry_tag >= JVM_CONSTANT_InternalMin && entry_tag <= JVM_CONSTANT_InternalMax) ||
+            (entry_tag == JVM_CONSTANT_Parameter) ||
+            (entry_tag == JVM_CONSTANT_Linkage))
            || (((tag & JVM_CONSTANT_QDescBit) != 0) && (entry_tag == JVM_CONSTANT_Class ||
                entry_tag == JVM_CONSTANT_UnresolvedClass || entry_tag == JVM_CONSTANT_UnresolvedClassInError
                || entry_tag == JVM_CONSTANT_ClassIndex))
